artiq/test/test_tools.py

```diff
@@ -0,0 +1,125 @@
+from contextlib import contextmanager
+import unittest
+from pathlib import Path
+import random
+import string
+import tempfile
+
+from artiq import tools
+
+
+# Helper to create temporary modules
+# Very simplified version of CPython's
+# Lib/test/test_importlib/util.py:create_modules
+@contextmanager
+def create_modules(*names):
+    mapping = {}
+    with tempfile.TemporaryDirectory() as temp_dir:
+        mapping[".root"] = Path(temp_dir)
+
+        for name in names:
+            file_path = Path(temp_dir) / f"{name}.py"
+            with file_path.open("w") as fp:
+                print(f"_MODULE_NAME = {name!r}", file=fp)
+            mapping[name] = file_path
+
+        yield mapping
+
+
+def random_chars(length=8):
+    return "".join(random.choices(string.ascii_letters, k=length))
```

Comment:
Probably not an issue here but it's generally good to avoid randomness in unit tests, to avoid non-reproducible results.